mm2src/gui_storage/src/account/storage/sqlite_storage.rs

```diff
@@ -0,0 +1,580 @@
+        query
+            .field(account_table::ACCOUNT_TYPE)?
+            .field(account_table::ACCOUNT_IDX)?
+            .field(account_table::DEVICE_PUBKEY)?
+            .field(account_table::NAME)?
+            .field(account_table::DESCRIPTION)?
+            .field(account_table::BALANCE_USD)?;
+        let accounts = query
+            .query(account_from_row)?
+            .into_iter()
+            .map(|account| (account.account_id.clone(), account))
+            .collect();
+        Ok(accounts)
+    }
+
+    fn account_exists(conn: &Connection, account_id: &AccountId) -> AccountStorageResult<bool> {
+        let mut query = SqlQuery::select_from(conn, account_table::TABLE_NAME)?;
+        query.count(account_table::NAME)?;
+
+        let (account_type, account_idx, device_pubkey) = account_id.to_sql_tuple();
+        query
+            .and_where_eq(account_table::ACCOUNT_TYPE, account_type)?
+            .and_where_eq(account_table::ACCOUNT_IDX, account_idx)?
+            .and_where_eq_param(account_table::DEVICE_PUBKEY, device_pubkey)?;
+
+        let accounts = query
+            .query_single_row(count_from_row)?
+            .or_mm_err(|| AccountStorageError::Internal("'count' should have returned one row exactly".to_string()))?;
+
+        Ok(accounts > 0)
+    }
+
+    fn upload_account(conn: &Connection, account: AccountInfo) -> AccountStorageResult<()> {
+        let mut sql_insert = SqlInsert::new(conn, account_table::TABLE_NAME);
+
+        let (account_type, account_idx, device_pubkey) = account.account_id.to_sql_tuple();
+        sql_insert
+            .column(account_table::ACCOUNT_TYPE, account_type)?
+            .column(account_table::ACCOUNT_IDX, account_idx)?
+            .column_param(account_table::DEVICE_PUBKEY, device_pubkey)?
+            .column_param(account_table::NAME, account.name)?
+            .column_param(account_table::DESCRIPTION, account.description)?
+            .column_param(account_table::BALANCE_USD, account.balance_usd.to_string())?;
+        sql_insert.insert()?;
+        Ok(())
+    }
+
+    fn delete_account(conn: &mut Connection, account_id: AccountId) -> AccountStorageResult<()> {
+        let transaction = conn.transaction()?;
+
+        // Check if the account can be enabled.
+        if let Some(enabled_account_id) = account_id.try_to_enabled() {
+            // Delete the account from `enabled_account_table` if **only** it's enabled.
+            let (account_type, account_idx) = enabled_account_id.to_sql_pair();
+            let mut sql_delete_enabled = SqlDelete::new(&transaction, enabled_account_table::TABLE_NAME)?;
+            sql_delete_enabled
+                .and_where_eq(enabled_account_table::ACCOUNT_TYPE, account_type)?
+                .and_where_eq(enabled_account_table::ACCOUNT_IDX, account_idx)?;
+            sql_delete_enabled.delete()?;
+        }
+
+        let (account_type, account_idx, device_pubkey) = account_id.to_sql_tuple();
```

Comment:
This can be moved above `if let Some(enabled_account_id) = account_id.try_to_enabled() {` and use `account_type`, `account_idx` there instead of calling `let (account_type, account_idx) = enabled_account_id.to_sql_pair();`

Reply:
I've done it for the reason that it will help to avoid bugs in the future, if `AccountId::to_sql_tuple()` and `EnabledAccountId::to_sql_pair()` calculates `account_type` and `account_idx` differently.
I actually don't think that it's the case, but I thought it wouldn't be superfluous to handle them separately.

Reply:
> Do you think this is the reason to leave it as it is? 🙂

Yeah, sure. Sine you have a good reason for this.

Reply:
I fixed this comment during last refactoring.